### ISSUE TITLE
Add wiki entries 

### ISSUE THREAD

**Issue by dp247** [robbrad/UKBinCollectionData#100]
Add instructions for the following councils:

- [x] York
- [x] East Riding
- [x] Windsor and Maidenhead (once complete)
- [x] Castlepoint

**Comment by dp247** [comment]
Kinda related to #55 I guess

**Comment by dp247** [comment]
Pretty sure there's more councils to be added too

**Comment by robbrad** [comment]
There are 57 Councils supported (albeit some are broken)

**Comment by OliverCullimore** [comment]
@robbrad @dp247 Just a thought on this, if we created a wiki/docs folder in the main repo using https://github.com/marketplace/actions/github-wiki-action the wiki could pull from that allowing it to be updated as part of pull requests for adding councils?

**Comment by robbrad** [comment]
@OliverCullimore - regarding the change to the workflow trigger for the gh-pages (Behave reports) and the run of the integration tests - whats the thinking here (I was setting up the env to take my email) see the link below and on commit the workflow didnt trigger

https://github.com/robbrad/UKBinCollectionData/blob/master/.github/workflows/wiki.yml#L19

Is it to reduce workflow runs? or is it to force wiki updates ? - my perspective is here for the testing at least is we always want unit tests and intergration tests to run on every commit/merge incase it breaks a council

Now with that said its not perfect - its running on every branch creation etc at the moment and on master when in reality merge to master prob isnt required - eg if the branch run passed - but this might be a wider discussion point

**Comment by robbrad** [comment]
Sorry only just realised the !wiki/** is an exclusion

Strangely the workflow didnt run for this PR https://github.com/robbrad/UKBinCollectionData/pull/216

**Comment by robbrad** [comment]
I think we need to use paths-ignore

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

**Comment by OliverCullimore** [comment]
@robbrad - It was an attempt to prevent the behave and codecov actions running if someone just committed changes to the wiki md files and not any actual code.

Then also an attempt to only run the deploy wiki action only if wiki md files have been changed in a commit.

Ideally we'd only run the wiki action on a push/merge rather than on creation etc of pull requests also to reduce unnecessary runs of it further.

As I mentioned I wasn't able to test the actions and just created/amended as per the docs that seemed applicable so they could need some tweaks to achieve the intended behaviour above.

**Comment by robbrad** [comment]
No worries @OliverCullimore - I worked out it needed paths-igore so I learnt something new today !

It has also forced me to optimise the workflows which is a good thing

Im just making the last few changes now

**Comment by OliverCullimore** [comment]
@robbrad ref your comment on #213 
> This is great @OliverCullimore - could one idea be to suck up the wiki entries from the top of the parser file? Just an idea

Would you be able to elaborate on this?

Do you mean adding comments with the details for the wiki to the top of the python files in the councils folder which we then build the Councils.md wiki page from?

I did originally think about us having a JSON file within the repository root or the uk_bin_collection folder which would define all the councils, their notes for the wiki and any details needed for their integration tests which could supersede/generate the input.json and validate_council_outputs.feature files as well as automatically generate the Councils wiki page for us.

But didn't want to suggest such a widespread change that could involve work on how all the testing is set up.

E.g.

Councils.json
```json
{
    "CheshireEastCouncil": {
        "url": "https://online.cheshireeast.gov.uk/MyCollectionDay/SearchByAjax/GetBartecJobList?uprn=100012791226&onelineaddress=3%20COBBLERS%20YARD,%20SK9%207DZ&_=1621149987573",
        "uprn": "100012791226",
        "wiki_name": "Cheshire East Council",
        "wiki_command_url_override": "https://online.cheshireeast.gov.uk/MyCollectionDay/SearchByAjax/GetBartecJobList?uprn=XXXXXXXX&onelineaddress=XXXXXXXX&_=1621149987573",
        "wiki_note": "Both the UPRN and a one-line address are passed in the URL, which needs to be wrapped in double quotes.\nThe one-line address is made up of the house number, street name and postcode.\nUse the form [here](https://online.cheshireeast.gov.uk/mycollectionday/) to find them, then take the first line and postcode and replace all spaces with `%20`."
    }
}
```

Which we'd then be able to use to generate wiki entries for us with something along the lines of:

```python
councils_wiki = ""
        json_data = json.loads(councils_file)
        for council, council_details in json_data:
            if council != "" and council_details.get("wiki_name", "") != "":
                command = " " + council_details.get("wiki_command_url_override", council_details.get("url", ""))
                if "uprn" in council_details:
                    command += " -u XXXXXXXX"
                if "postcode" in council_details:
                    command += " -p \"XXXX XXX\""
                if "house_number" in council_details:
                    command += " -n XX"
                if "SKIP_GET_URL" in council_details:
                    command += " --skip_get_url"
                councils_wiki += "\n ---\n"
                councils_wiki += " ### " + council_details["wiki_name"] + "\n"
                councils_wiki += " ```commandline\n"
                councils_wiki += " python collect_data.py " + council + command + "\n"
                councils_wiki += " ```\n"
                if council_details.get("wiki_note", "") != "":
                    councils_wiki += council_details.get("wiki_note", "") + "\n"
```

At the same time, I also questioned if we could generate the schema files from the JSON outputs in real-time during tests possibly in the load_schema_file() function or as a workflow, but wasn't sure what would be involved with this or if it would prevent the tests being as effective or anything?

**Comment by OliverCullimore** [comment]
> No worries @OliverCullimore - I worked out it needed paths-ignore so I learnt something new today !
> 
> It has also forced me to optimise the workflows which is a good thing
> 
> Im just making the last few changes now

@robbrad ref this again I noticed https://github.com/robbrad/UKBinCollectionData/pull/231#issuecomment-1445484549_ so it looks like we may need to exclude `*.md` files as well/instead as the change to CONTRIBUTING.md caused the action to still run?

**Comment by robbrad** [comment]
> @robbrad ref your comment on #213
> 
> > This is great @OliverCullimore - could one idea be to suck up the wiki entries from the top of the parser file? Just an idea
> 
> Would you be able to elaborate on this?
> 
> Do you mean adding comments with the details for the wiki to the top of the python files in the councils folder which we then build the Councils.md wiki page from?
> 
> I did originally think about us having a JSON file within the repository root or the uk_bin_collection folder which would define all the councils, their notes for the wiki and any details needed for their integration tests which could supersede/generate the input.json and validate_council_outputs.feature files as well as automatically generate the Councils wiki page for us.
> 
> But didn't want to suggest such a widespread change that could involve work on how all the testing is set up.
> 
> E.g.
> 
> Councils.json
> 
> ```json
> {
>     "CheshireEastCouncil": {
>         "url": "https://online.cheshireeast.gov.uk/MyCollectionDay/SearchByAjax/GetBartecJobList?uprn=100012791226&onelineaddress=3%20COBBLERS%20YARD,%20SK9%207DZ&_=1621149987573",
>         "uprn": "100012791226",
>         "wiki_name": "Cheshire East Council",
>         "wiki_command_url_override": "https://online.cheshireeast.gov.uk/MyCollectionDay/SearchByAjax/GetBartecJobList?uprn=XXXXXXXX&onelineaddress=XXXXXXXX&_=1621149987573",
>         "wiki_note": "Note: Both the UPRN and a one-line address are passed in the URL, which needs to be wrapped in double quotes.\nThe one-line address is made up of the house number, street name and postcode.\nUse the form [here](https://online.cheshireeast.gov.uk/mycollectionday/) to find them, then take the first line and postcode and replace all spaces with `%20`."
>     }
> }
> ```
> 
> Which we'd then be able to use to generate wiki entries for us with something along the lines of:
> 
> ```python
> wiki_content = "\n\n" +
> "As a reminder, most scripts only need a module name and a URL to run, but others need more parameters depending on how the data is scraped..." +
> "\n\n## Contents"
> wiki_entries = ""
> json_data = json.loads(councils_file)
> for council, council_details in json_data:
>     if council != "" and council_details.get("wiki_name", "") != "":
>         wiki_content += "\n- [" + council_details["wiki_name"] + "](#" + council.lower().replace(" ", "-") + ")"
>         command = council_details.get("wiki_command_url_override", council_details.get("url", ""))
>         if "uprn" in council_details:
>             command += " -u XXXXXXXX"
>         if "postcode" in council_details:
>             command += " -p \"XXXX XXX\""
>         if "house_number" in council_details:
>             command += " -n XX"
>         if "SKIP_GET_URL" in council_details:
>             command += " -s true"
>         wiki_entries += "\n ---\n"
>         wiki_entries += " ### " + council_details["wiki_name"] + "\n"
>         wiki_entries += " ```commandline\n"
>         wiki_entries += " python collect_data.py " + council + " " command + "\n"
>         wiki_entries += " ```\n"
>         if council_details.get("wiki_note", "") != "":
>             wiki_entries += council_details.get("wiki_note", "") + "\n"
> wiki_content += "\n" + wiki_entries
> ```
> 
> At the same time, I also questioned if we could generate the schema files from the JSON outputs in real-time during tests possibly in the load_schema_file() function or as a workflow, but wasn't sure what would be involved with this or if it would prevent the tests being as effective or anything?

Yep you nailed it - the entries in the imput.json wont break the tests

On the schema generation - we want that to be a speroate process - only because sometimes the councils spit out { bins: [] } which is valid json and would prob have a vaild schema - but the test result would so success. Hence why we want to have a prescriptive "what we expect" vs "what we get"

But to keep the parity once its achieved - a small pipeline to validate input.json vs feature file vs schemas vs parsers would be useful

**Comment by robbrad** [comment]
> > No worries @OliverCullimore - I worked out it needed paths-ignore so I learnt something new today !
> > It has also forced me to optimise the workflows which is a good thing
> > Im just making the last few changes now
> 
> @robbrad ref this again I noticed [#231 (comment)](https://github.com/robbrad/UKBinCollectionData/pull/231#issuecomment-1445484549)_ so it looks like we may need to exclude `*.md` files as well/instead as the change to CONTRIBUTING.md caused the action to still run?

This should do it https://github.com/robbrad/UKBinCollectionData/pull/234 - can you or @dp247 review?

I have gotten bad at skipping the reviewer recently and we might wanna tighten that up as parity is achieved

![image](https://user-images.githubusercontent.com/6122503/221683245-7ae62fe2-6ecf-4aa4-a83a-746efc2f5d3a.png)

**Comment by OliverCullimore** [comment]
> > @robbrad ref your comment on #213
> > 
> > > This is great @OliverCullimore - could one idea be to suck up the wiki entries from the top of the parser file? Just an idea
> > 
> > Would you be able to elaborate on this?
> > 
> > Do you mean adding comments with the details for the wiki to the top of the python files in the councils folder which we then build the Councils.md wiki page from?
> > 
> > I did originally think about us having a JSON file within the repository root or the uk_bin_collection folder which would define all the councils, their notes for the wiki and any details needed for their integration tests which could supersede/generate the input.json and validate_council_outputs.feature files as well as automatically generate the Councils wiki page for us.
> > 
> > But didn't want to suggest such a widespread change that could involve work on how all the testing is set up.
> > 
> > E.g.
> > 
> > Councils.json
> > 
> > ```json
> > {
> >     "CheshireEastCouncil": {
> >         "url": "https://online.cheshireeast.gov.uk/MyCollectionDay/SearchByAjax/GetBartecJobList?uprn=100012791226&onelineaddress=3%20COBBLERS%20YARD,%20SK9%207DZ&_=1621149987573",
> >         "uprn": "100012791226",
> >         "wiki_name": "Cheshire East Council",
> >         "wiki_command_url_override": "https://online.cheshireeast.gov.uk/MyCollectionDay/SearchByAjax/GetBartecJobList?uprn=XXXXXXXX&onelineaddress=XXXXXXXX&_=1621149987573",
> >         "wiki_note": "Note: Both the UPRN and a one-line address are passed in the URL, which needs to be wrapped in double quotes.\nThe one-line address is made up of the house number, street name and postcode.\nUse the form [here](https://online.cheshireeast.gov.uk/mycollectionday/) to find them, then take the first line and postcode and replace all spaces with `%20`."
> >     }
> > }
> > ```
> > 
> > Which we'd then be able to use to generate wiki entries for us with something along the lines of:
> > 
> > ```python
> > wiki_content = "\n\n" +
> > "As a reminder, most scripts only need a module name and a URL to run, but others need more parameters depending on how the data is scraped..." +
> > "\n\n## Contents"
> > wiki_entries = ""
> > json_data = json.loads(councils_file)
> > for council, council_details in json_data:
> >     if council != "" and council_details.get("wiki_name", "") != "":
> >         wiki_content += "\n- [" + council_details["wiki_name"] + "](#" + council.lower().replace(" ", "-") + ")"
> >         command = council_details.get("wiki_command_url_override", council_details.get("url", ""))
> >         if "uprn" in council_details:
> >             command += " -u XXXXXXXX"
> >         if "postcode" in council_details:
> >             command += " -p \"XXXX XXX\""
> >         if "house_number" in council_details:
> >             command += " -n XX"
> >         if "SKIP_GET_URL" in council_details:
> >             command += " -s true"
> >         wiki_entries += "\n ---\n"
> >         wiki_entries += " ### " + council_details["wiki_name"] + "\n"
> >         wiki_entries += " ```commandline\n"
> >         wiki_entries += " python collect_data.py " + council + " " command + "\n"
> >         wiki_entries += " ```\n"
> >         if council_details.get("wiki_note", "") != "":
> >             wiki_entries += council_details.get("wiki_note", "") + "\n"
> > wiki_content += "\n" + wiki_entries
> > ```
> > 
> > At the same time, I also questioned if we could generate the schema files from the JSON outputs in real-time during tests possibly in the load_schema_file() function or as a workflow, but wasn't sure what would be involved with this or if it would prevent the tests being as effective or anything?
> 
> Yep you nailed it - the entries in the imput.json wont break the tests
> 
> On the schema generation - we want that to be a speroate process - only because sometimes the councils spit out { bins: [] } which is valid json and would prob have a vaild schema - but the test result would so success. Hence why we want to have a prescriptive "what we expect" vs "what we get"
> 
> But to keep the parity once its achieved - a small pipeline to validate input.json vs feature file vs schemas vs parsers would be useful

Sorry on the schema generation by JSON outputs I was referring to the outputs/CouncilName.json files which are manually populated and committed. I was thinking it may be possible for the tests to grab the applicable one for the council it's testing and generate a schema from it to then validate against the test's JSON result. Hope that makes sense?

On the wiki generation, I was thinking this might be best implemented in python as generate_wiki.py somewhere in the project which a workflow then runs? Maybe within the wiki folder and exclude it from the wiki deployment (by adding `EXCLUDED_FILES: "generate_wiki.py"` to the wiki action env list)?

**Comment by robbrad** [comment]
@OliverCullimore got you

Re: outputs
Yes that's possible - they actually don't serve any purpose other than to help people by being able to see what the output looks like.

My vote would be a separate generator that runs weekly and commits back to the repo or just a generator that we use on commit when required ( or if we do a parity pipeline this could be a job for that)

Re: Wiki generator 
Yes this sounds like a good approach (I'm not wanting to try to over complicate the process, but am trying to make it easy for people to keep the wiki updated)

**Comment by OliverCullimore** [comment]
@robbrad I'll look to submit a PR for the wiki changes when I can.

Re: schema
Now you mention something we can run on commit, we could achieve this with the addition of a -dev flag for collect_data.py which I'm now thinking could create/update the output folder JSON matching the council name you're running for you and also create a schema file to match? This would just save us creating/updating them manually and we'd commit them as per usual.

**Comment by robbrad** [comment]
I haven't found a decent schema creation tool yet other than the website linked in the contributing.md but if it can be programmatically done then I'm all for that too!

**Comment by OliverCullimore** [comment]
@robbrad it doesn't look like #239 went totally to plan as I told it to push the wiki updates the master repo which requires approval so it failed, are we able to add an exception that allows it only to push straight to the master branch without approval?

Otherwise, I'm thinking we could remove the Councils.md file from the repository as well as my commit step in the workflow as it should still deploy the generated Councils.md file to the wiki with the other Markdown files?

**Comment by robbrad** [comment]
I thin an exception might be the way to go

**Comment by OliverCullimore** [comment]
@robbrad @dp247 besides fixing the Wiki action permissions to allow it to commit the Councils.md file so when you clone/browse the repo you're seeing the up to date version, I think this one's complete now?

@robbrad have we made any progress on adding the exception? I did some Googling and it didn't sound too promising and sounds like a common issue people run into when trying to retain branch protection rules but allowing automation. We can resort to it creating a pull request we just merge if not (we'd likely also need some more trigger conditions to prevent those merges creating more PR's though).
